### PR TITLE
Fix long playlists (100+ videos)

### PR DESCRIPTION
**Pull request by SamantazFox**
This should fix issue #1883.
I've also made some changes in `src/invidious/channels.cr` because both channel and playlist videos require a call to `[YT]/youtubei/v1/browse`, so I made a common wrapper (it will ease future updates if the key or client version strings change).

### CODE REVIEW

**Comment by SamantazFox** [comment]
```
Failures:

  1) Helper #produce_playlist_url correctly produces url for requesting index `x` of a playlist
     Failure/Error: produce_playlist_continuation("UCCla9fZca4I7KagBtgRGnOw", 200).should eq("4qmFsgJLEhpWTFVVQ2xhOWZaY2E0STdLYWdCdGdSR25PdxoSQ0FKNkIxQlVPa05OWjBJJTNEmgIYVVVDbGE5ZlpjYTRJN0thZ0J0Z1JHbk93")

       Expected: "4qmFsgJLEhpWTFVVQ2xhOWZaY2E0STdLYWdCdGdSR25PdxoSQ0FKNkIxQlVPa05OWjBJJTNEmgIYVVVDbGE5ZlpjYTRJN0thZ0J0Z1JHbk93"
            got: "4qmFsgJLEhpWTFVVQ2xhOWZaY2E0STdLYWdCdGdSR25PdxoSQ0FGNkIxQlVPa05OWjBJJTNEmgIYVVVDbGE5ZlpjYTRJN0thZ0J0Z1JHbk93"

     # spec/helpers_spec.cr:54

Finished in 6.26 milliseconds
13 examples, 1 failures, 0 errors, 0 pending
``` 

One of the `produce_playlist_continuation` tests (see above) fail because:

1. I didn't bother implementing a request counter increment in the playlist continuation token (youtube seems to not care about it atm).
2. I copy/pasted reference tokens from actual youtube website tokens, whose counter are incremented

**Comment by TheFrenchGhosty** [comment]
What are the differences with #1909 ? 

PS: CI must be fixed.

**Comment by SamantazFox** [comment]
> What are the differences with #1909 ?
All playlists mentonned there work, being 300 or 5k videos ones and the continuation token was fixed too (so we can have pages >= 2)

> PS: CI must be fixed.
Yup, I know. I'll go on for the "dummy request index" (option 1), then, so if YT does take that into account in the future, that won't be a problem.


PS: Also I don't care about the bounty.

**Comment by TheFrenchGhosty** [comment]
>All playlists mentonned there work, being 300 or 5k videos ones and the continuation token was fixed too (so we can have pages >= 2)

Well, this is great then!

>Yup, I know. I'll go on for the "dummy request index" (option 1), then, so if YT does take that into account in the future, that won't be a problem.

Alright, I'll wait for this to be done before I ask saltycrys to review.

>PS: Also I don't care about the bounty.

You're sure? You deserve it. If you really don't want it, we can send it to someone else you choose if you want.

**Comment by SamantazFox** [comment]
> Alright, I'll wait for this to be done before I ask saltycrys to review.

It's fixed, all CI tests pass now :)

> You're sure? You deserve it. If you really don't want it, we can send it to someone else you choose if you want.

I don't really like cryptos, tbh. I know someone who was *very* excited about that bounty :P

File: src/invidious/playlists.cr [hunk 1]
```diff
@@ -307,23 +307,32 @@ def subscribe_playlist(db, user, playlist)
...
+  # Emulate a "request counter" increment, to make perfectly valid
+  # ctokens, even if at the time of writing, it's ignored by youtube.
+  request_count = (index / 100).to_i64 || 1_i64
+
   data = {"1:varint" => index.to_i64}
     .try { |i| Protodec::Any.cast_json(i) }
     .try { |i| Protodec::Any.from_json(i) }
     .try { |i| Base64.urlsafe_encode(i, padding: false) }
 
+  data_wrapper = {"1:varint" => request_count, "15:string" => "PT:#{data}"}
```
Comment (tenpura-shrimp) [inline src/invidious/playlists.cr:325]: Why did you change to this instead of using ```base64```?  That was made to do exactly this, though its not exactly documented.

https://github.com/omarroth/protodec/blob/master/src/protodec/utils.cr#L198

File: src/invidious/playlists.cr [hunk 2]
```diff
@@ -307,23 +307,32 @@ def subscribe_playlist(db, user, playlist)
...
+
   object = {
     "80226972:embedded" => {
-      "2:string" => plid,
-      "3:base64" => {
-        "15:string" => "PT:#{data}",
-      },
+      "2:string"  => plid,
+      "3:string"  => data_wrapper,
+      "35:string" => id,
```
Comment (tenpura-shrimp) [inline src/invidious/playlists.cr:335]: I'm assuming, but not completely sure that this additional field is why other PR wasn't working

File: src/invidious/playlists.cr (hunk 1 above)
Comment (SamantazFox) [reply src/invidious/playlists.cr:325]: I did that because "base64" doesn't allow the data to be URL-encoded (i.e with "=" replaced with "%3D").

I did not try to understand why it was done this way, I just tried to reproduce the YT token generated by youtube. That was the only way that worked.

File: src/invidious/playlists.cr (hunk 2 above)
Comment (SamantazFox) [reply src/invidious/playlists.cr:335]: Yep, definitely. Along with the `youtubei/v1/browse_ajax` vs `youtubei/v1/browse` API endpoint.

File: src/invidious/channels.cr [hunk 3]
```diff
@@ -938,34 +939,18 @@ def get_about_info(ucid, locale)
 end
 
 def get_channel_videos_response(ucid, page = 1, auto_generated = nil, sort_by = "newest", youtubei_browse = true)
```
Comment (saltycrys) [inline src/invidious/channels.cr:941]: The `youtubei_browse` parameter can be removed.

**Review by saltycrys — APPROVED** [review]

File: src/invidious/channels.cr (hunk 3 above)
Comment (SamantazFox) [reply src/invidious/channels.cr]: Yes, forgot to. Fixed in 3e88b72.

**Comment by TheFrenchGhosty** [comment]
I'll surely test this PR tonight, if everything works I'll merge it.

@SamantazFox what is your decision about the reward?

**Comment by SamantazFox** [comment]
> @SamantazFox what is your decision about the reward?

Give it to @tenpura-shrimp ^^ They seem to need it more than me!

**Comment by TheFrenchGhosty** [comment]
@SamantazFox Alright! Will do.

**Comment by tenpura-shrimp** [comment]
I won't take this one because it's mostly fox's work here. You can discuss further or save it for a future bounty.

**Comment by TheFrenchGhosty** [comment]
Tested, this works almost perfect, except for one "small" issue:

https://www.youtube.com/playlist?list=PL_oFlvgqkrjUVQwiiE3F3k3voF4tjXeP0 when you are on the last page (page 11), you can click next page, and it will go to page 12 (that doesn't exist), and an "empty" page will be shown.

Other than that, it's perfect.

**Comment by SamantazFox** [comment]
@TheFrenchGhosty Ah, yeah, I didn't fix the HTML template nor the paging code.

**Comment by SamantazFox** [comment]
Here we go. That's fixed :)

**Review by saltycrys — APPROVED** [review]

**Comment by TheFrenchGhosty** [comment]
>Oh, and BTW, I don't have page 11. It stops at page 10 for me. Maybe some videos are geo-blocked?

I test that on my instance on https://pussthecat.org so it's in Germany. It's possible that it's geo-bocked in France yeah.

**Comment by TheFrenchGhosty** [comment]
Tested: works perfectly!

The fact that when going over the page number you get redirect to the last page ( https://github.com/iv-org/invidious/blob/23e5b6ba72c3c39df97c4fd21980997b9a30e303/src/invidious/routes/playlists.cr#L439 ) is really cool too.

**Comment by TheFrenchGhosty** [comment]
Merged.

Thanks for the review @saltycrys 

Thanks @SamantazFox and @tenpura-shrimp 

If you really don't want the reward we'll surely just use this money for another bounty.

**Comment by SamantazFox** [comment]
> The fact that when going over the page number you get redirect to the last page is really cool too.

glad you like it ^^

**Comment by TheFrenchGhosty** [comment]
@SamantazFox I mean, it's a clever way to limit it without displaying an "error" page, and actually useful.

**Comment by TheFrenchGhosty** [comment]
@SamantazFox I think this PR is breaking comments by adding  ```?format=html&hl=en-US&thin_mode=false``` at the end of the API url.
